contracts/SessionBatchInvoker.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {InvokerBase} from "./InvokerBase.sol";
+import {ReentrancyGuard} from "./utils/ReentrancyGuard.sol";
+import {Ownable} from "./utils/Ownable.sol"; // TODO: TBD.
+
+/**
+ * @title Session Batch Invoker
+ * @author Paul O'Leary <poleary@polygon.technology>, Zero Ekkusu <zeroekkusu.eth>
+ * @notice An EIP-3074, session based contract that can send one or more arbitrary transactions in the context of an Externally
+ *  Owned Address (EOA), by using `AUTH` and `AUTHCALL`. See https://github.com/0xPolygon/account-abstraction-invoker for more
+ *  information.
+ */
+contract SessionBatchInvoker is InvokerBase, ReentrancyGuard, Ownable {
+    struct Transaction {
+        address from;
+        SessionToken token;
+        TransactionPayload[] payloads;
+    }
+
+    struct SessionToken {
+        address delegate; // token is bound to specific (trusted) delegate
+        uint256 expiration; // time of token expiration
+        bool revocable;
+    }
+
+    struct WhitelistChange {
+        address to;
+        bool allowed;
+    }
+
+    struct WhitelistUpdate {
+        WhitelistChange[] changes;
+        uint256 applicableFrom;
+    }
+
+    bytes32 private constant TRANSACTION_TYPE = keccak256(
+        "Transaction(address from,SessionToken token,TransactionPayload[] payloads)SessionToken(address delegate,uint256 expiration,bool revocable)TransactionPayload(address to,uint256 value,uint256 gasLimit,bytes data)"
+    );
+
+    bytes32 private constant SESSION_TOKEN_TYPE =
+        keccak256("SessionToken(address delegate,uint256 expiration,bool revocable)");
+
+    uint256 private immutable maxSessionTokenDuration;
```

Comment:
I think this should likely be mutable. It seems reasonable to me that a service provider could change their mind about the optimal duration.